### DISCUSSION TITLE
Is pulse agendaJs complient? Will it work with CosmosDB

### DISCUSSION THREAD

**Discussion by cemusta** [pulsecron/pulse#36] [General]
Hey,

Can we drop and use pulse with an existing AgendaJs database? I noticed there's a breaking change in Job.Define(). I was wondering if I could just drop the old AgendaJs and start using the Pulse?

**Comment by code-xhyun** [comment]
Of course, you'll need to run additional tests, but generally, Pulse's data structure is as follows, which isn't significantly different from the version of Agenda I've checked. However, as you know, Agenda has multiple versions, so please ensure thorough testing. If you encounter any issues during testing, let me know, and I'll address them immediately.




```
//pulse data example
{
  "_id": {
    "$oid": "664ae82720d4b2e571170a31"
  },
  "name": "send email",
  "data": {},
  "priority": 0,
  "shouldSaveResult": false,
  "attempts": 0,
  "backoff": null,
  "type": "normal",
  "nextRunAt": {
    "$date": "2024-05-21T08:50:22.641Z"
  },
  "repeatInterval": "1 minute",
  "repeatTimezone": null,
  "startDate": null,
  "endDate": null,
  "skipDays": null,
  "lastModifiedBy": null,
  "lockedAt": null,
  "lastRunAt": {
    "$date": "2024-05-21T08:49:22.641Z"
  },
  "runCount": 134,
  "finishedCount": 134,
  "lastFinishedAt": {
    "$date": "2024-05-21T08:49:22.642Z"
  }
}

```

**Comment by cemusta** [comment]
Hi, 

I did an early test, I'm getting **Unsupported BSON version, bson types must be from bson 6.x.x** errors atm. 

However, when I run some old jobs by manually setting `nextRunAt` by hand they run again successfully.

The existing system used the latest agendaJs (5.0) with Azure CosmosDb.

I just added this bold part below so pulse will use the old collection.

```ts
    this.agenda = new Pulse({
      db: {
        address: this.configService.get<string>('DB_CONNECTION'),
        collection: 'agendaJobs', // added this
      },
      processEvery: '5 seconds',
      sort: { nextRunAt: 1 },
    });
 ```
 
 Also I reorder the Job.define() parameters so configs become 3rd parameter from 2nd one. After this, there was no build problem.

**Reply by code-xhyun** [reply]
That's great! Is there anything else I can help you with?

**Reply by cemusta** [reply]
**there's a problem with Azure Cosmosdb**, should I create a new discussion for it?

Azure CosmosDB is mongoDb compliant db but its locked at mongo version 4x.

```ts
await this.pulse.now('testJob', {
          data
        });
```

While creating a new job, pulse is throwing this error: **Unsupported BSON version, bson types must be from bson 6.x.x**

**Reply by code-xhyun** [reply]
Currently, only the latest version of mongoDB is supported

**Reply by cemusta** [reply]
Is pulse using some new mongo v5+ features if not maybe its possible to make it work with v4 too?

**Reply by code-xhyun** [reply]
Sorry, We currently do not plan to support version 4x.
